### PR TITLE
Display child modules nested under parents

### PR DESCRIPTION
**Pull request by esoergel**
image

This PR displays child modules nested below their parents.  Rearrangements preserve the nested structure, so moving a parent also moves the children.  Child modules can be rearranged under the same parent, but not promoted to top level or moved to a new parent.

Commit by commit will make the most sense.
@emord @snopoke

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 1]
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
...
-        });
-        $('.sort-action').hide();
+        if (modulesWereReordered()) {
+            promptToSaveOrdering();
+        } else {
+            $('.sortable').each(function () {
+                initSortable($(this));
+            });
+            $('.sort-action').hide();
+        }
```
Comment (esoergel) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: The promptToSaveOrdering will be dismissed if the user clicks outside the modal, but they won't be able to reorder the modules, since they're not initialized here.

As per the commit message, I don't expect this situation to arise often, but it could, since we currently have applications with child modules that don't follow their parent.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 2]
```diff
@@ -265,8 +265,13 @@ hqDefine('app_manager/js/app_manager', function () {
...
                 var parent = modulesByUid[$(childModule).data('rootmoduleid')];
                 if (!parent) {
-                    // This child module is orphaned, throw it at the end
+                    // This child module is orphaned, throw it at the end and alert user
                     $("ul.appnav-module").append(childModule);
+                    var moduleName = $(childModule).data('modulename'),
+                        alertText = _.template(gettext(
+                        'Menu "<%= moduleName %>" references a parent menu that no longer exists'
+                    ))({moduleName: moduleName});
+                    hqImport('hqwebapp/js/alert_user').alert_user(alertText, "danger");
```
Comment (esoergel) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: When they navigate to the module, there's also a warning on the page.

**Comment by esoergel** [comment]
This PR does force some projects to reorder their modules.  There are comparatively few projects impacted by this, but the solutions team will be reaching out to those projects to "fix" their ordering prior to releasing this feature.

File: corehq/apps/app_manager/templates/app_manager/partials/menu/module_link.html [hunk 3]
```diff
@@ -0,0 +1,109 @@
+{% load i18n %}
+{% load xforms_extras %}
+{% load hq_shared_tags %}
+
+{# For APP BUILDER PROTOTYPE use only #}
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/module_link.html]: this would be a nice comment to kill

File: corehq/apps/app_manager/models.py [hunk 4]
```diff
@@ -5410,14 +5410,14 @@ def rearrange_modules(self, i, j):
             raise RearrangeError()
         self.modules = modules
 
-    def rearrange_forms(self, to_module_id, from_module_id, i, j):
+    def rearrange_forms(self, to_module_uid, from_module_uid, i, j):
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: hooray

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 5]
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
...
             $sortable.sortable(options);
         }
-
         function updateAfterMove(e, ui, $sortable) {
             // because the event is triggered on both sortables when moving between one sortable list to
             // another, do a check to see if this is the sortable list we're moving the item to
-            if ($sortable.find(ui.item).length < 1) {
-                return;
-            }
+            if ($sortable.find(ui.item).length < 1) { return; }
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: I abhor this style...maybe we can argue about it in book club.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 6]
```diff
@@ -232,12 +232,159 @@ hqDefine('app_manager/js/app_manager', function () {
...
+
+            var toModuleUid = $sortable.parents('.edit-module-li').data('uid'),
+                fromModuleUid = ui.item.data('moduleuid'),
+                sortingForms = $sortable.hasClass('sortable-forms'),
+                movingFormToNewModule = sortingForms && toModuleUid !== fromModuleUid;
+
+            var from, to;
+            if (movingFormToNewModule) {
+                [from, to] = calculateMoveFormToNewModule($sortable, ui, toModuleUid);
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Huh, I haven't seen that style in js. I like the style from a coding perspective, but I'm opposed to parse errors, since they can make it harder to integrate with other tools that expect correct syntax. We've had issues with the uglifier failing on js syntax that's illegal but accepted by browsers, specifically having a trailing comma in a list of function args.

I'm curious, does the uglier work properly on this? You can test by running `build_requirejs` locally (make sure to `rm -rf staticfiles` and run `collectstatic` and `compilejsi18n` first).

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 7]
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
...
                 hqImport("app_manager/js/menu").setPublishStatus(true);
             }
         }
-
+        function calculateMoveFormToNewModule($sortable, ui, toModuleUid) {
+            var from = -1, to = -1;
+            $sortable.children().not('.sort-disabled').each(function (i) {
+                if ($(this).data('moduleuid') !== toModuleUid) {
+                    to = i;
+                    from = parseInt(ui.item.data('index'), 10);
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Is the `parseInt` necessary? Usually jquery's `data` will parse numbers out automatically.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 8]
```diff
@@ -279,83 +265,106 @@ hqDefine('app_manager/js/app_manager', function () {
...
+                    } else {
+                        to = i;
+                        from = index;
+                        return false;
+                    }
+                }
+            });
+            return [from, to];
+        }
+        function saveDataToForm($form, from, to, fromModuleUid, toModuleUid, sortingForms) {
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Nitpick: this name is vague. `storeDestinationAndTarget`?

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 9]
```diff
@@ -247,7 +248,34 @@ hqDefine('app_manager/js/app_manager', function () {
...
             });
         }
-
+        function nestChildModules() {
+            var modulesByUid = {},
+                childModules = [];
+            $(".module").each(function (index, element) {
+                $(element).data('index', index);
+                modulesByUid[ $(element).data('uid') ] = element;
+                if ( $(element).data('rootmoduleid') ) {
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Should this be `rootmoduleuid`? I just went back to check this wouldn't run into issues with rootmoduleid being zero, rootmoduleuid would make it more obvious that isn't an issue.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 10]
```diff
@@ -247,7 +248,34 @@ hqDefine('app_manager/js/app_manager', function () {
...
+                    // This child module is orphaned, throw it at the end
+                    $("ul.appnav-module").append(childModule);
+                } else {
+                    addChildModuleToParent(childModule, parent);
+                }
+            });
+        }
+        function addChildModuleToParent(childModule, parent) {
+            var childList = $(parent).find("ul.child-modules");
+            if ( childList.length === 0 ) {
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Nitpick: Are you attached to the style of a space after opening paren/bracket and before closing paren/bracket? I feel like we don't use it much, at least in js.

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html [hunk 11]
```diff
@@ -12,11 +12,18 @@
     </li>
     {% with module as selected_module %}
         {% for module in app.get_modules %}
-            <li class="edit-module-li js-sorted-li {% if module.root_module_id %}child-module{% endif %}" data-index="{{ module.id }}" data-uid="{{ module.unique_id }}" data-indexvar="moduleid">
+            <li class="edit-module-li js-sorted-li module"
+                data-index="{{ module.id }}"
+                data-uid="{{ module.unique_id }}"
+                data-rootModuleId="{{ module.root_module_id|default:"" }}"
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: Nitpick: is the `default` necessary?

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 12]
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
     var _initMenuItemSorting = function () {
         nestChildModules();
         initDragHandles();
-        $('.sortable').each(function () {
-            initSortable($(this));
-        });
-        $('.sort-action').hide();
+        if (modulesWereReordered()) {
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: quality function name

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 13]
```diff
@@ -234,10 +234,14 @@ hqDefine('app_manager/js/app_manager', function () {
...
+            promptToSaveOrdering();
+        } else {
+            $('.sortable').each(function () {
+                initSortable($(this));
+            });
+            $('.sort-action').hide();
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Nitpick: It's a good habit to use `addClass('hide')` instead of `.hide()`, because the latter directly updates an inline style, which can screw with other styling.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: I think this is a good ui compromise.

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html [hunk 14]
```diff
@@ -67,3 +67,35 @@
...
+            </div>
+            <form action="{% url "reorder_child_modules" domain app.id %}"
+                  method="post">
+                {% csrf_token %}
+                <div class="modal-body">
+                    <p class="lead">
+                        {% trans "Your modules were reordered" %}
+                    </p>
+                    <p>
+                        Click below to move these child modules under their parents.
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: trans

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html [hunk 15]
```diff
@@ -67,3 +67,35 @@
...
+                    </p>
+                </div>
+                <div class="modal-footer">
+                    <button type="submit" class="disable-on-submit btn btn-danger">
+                        Rearrange
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: trans

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html [hunk 16]
```diff
@@ -67,3 +67,35 @@
...
+                <div class="modal-body">
+                    <p class="lead">
+                        {% trans "Your modules were reordered" %}
+                    </p>
+                    <p>
+                        Click below to move these child modules under their parents.
+                    </p>
+                </div>
+                <div class="modal-footer">
+                    <button type="submit" class="disable-on-submit btn btn-danger">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: Is this a dangerous action or is this what we want them to do?

Either way I'd add an explicit cancel button so people don't feel trapped.

File: corehq/apps/app_manager/views/apps.py [hunk 17]
```diff
@@ -902,6 +902,27 @@ def rearrange(request, domain, app_id, key):
...
+    )
+    orphaned_modules = []
+    for module in app.modules:
+        if module.root_module_id:
+            if module.root_module_id in modules_by_parent_id:
+                modules_by_parent_id[module.root_module_id].append(module)
+            else:
+                orphaned_modules.append(module)
+
+    app.modules = [m for modules in modules_by_parent_id.values() for m in modules]
```
Comment (orangejenny) [inline corehq/apps/app_manager/views/apps.py]: Nitpick: having `modules` twice inside the list comprehension is a little confusing.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: So this'll show a warning on every page in app manager, right? Given that there's a warning on the most relevant page and that there's a build error, I'm not convinced this is needed.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 10 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Nevermind, looks like stickler has the same complaint.

**Review by orangejenny — APPROVED** [review]
This is fabulous. My commentary is all minor.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 5 above)
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: which style?  doing this on one line, or short-circuiting early?

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 2 above)
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: That's a good point, I'll kill it.

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 7 above)
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Not sure, these were here already

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html (hunk 11 above)
Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: Yeah, otherwise it's `"None"`

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html (hunk 16 above)
Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: I am trying to force them to make that change, though I was also thinking about a "cancel", since popping this up every pageload and not allowing them to reorder modules is already a pretty aggressive way of forcing it.

File: corehq/apps/app_manager/views/apps.py (hunk 17 above)
Comment (esoergel) [reply corehq/apps/app_manager/views/apps.py]: what would you suggest?  I need to join the lists that are the dict values.  Could also do:
```python
list(itertools.chain(*modules_by_parent_id.values()))
```

**Review by proteusvacuum — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Dang, I'll just switch it to basic JS, how's this look?
```js
var move;
if ... {
    move = ...;
} else {
    move = ...;
}
var from = move[0],
    to = move[1]
```

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 13 above)
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Actually, do you know of any reason that's not just hidden in the html?  Why bother doing that in JS?

File: corehq/apps/app_manager/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/models.py]: Wow, good call, I was just updating this and totally assumed it'd be `rearrange_forms(from, to)`, but it's actually the reverse.

**Comment by esoergel** [comment]
Added a few commits in response to PR feedback, and updated the warning modal with some better copy.  

After looking a little more at the details of the rollout plan, I'm going to add in a feature flagged backwards compatibility shim so we don't have to migrate everyone (there are many dozens of copies of the ICDS apps floating around, and they have custom case tiles which have hardcoded module indices).

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 5 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: the single line

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js (hunk 13 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/app_manager.js]: Hmm, I don't know. I'd definitely favor doing it in html.

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html (hunk 16 above)
Comment (orangejenny) [reply corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: I'd mark it as `btn-primary` then.

File: corehq/apps/app_manager/views/apps.py (hunk 17 above)
Comment (orangejenny) [reply corehq/apps/app_manager/views/apps.py]: I think my real complaint is I don't quite get what this function is doing unless I go look at other commits for context. We're reordering modules, but why? What order are they going to end up in? Plus I find nested list comprehensions confusing which isn't helping. (In general when reviewing I waffle over whether to criticize nested comprehensions as difficult to read or just accept they're something I need to get more comfortable with.)

I think if the overall purpose of the function were clearer I wouldn't get hung up on this line. How about a function-level comment? Or a more descriptive function name?

File: corehq/apps/app_manager/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: Nice rename, any chance you'd mind also updating the order so it *is* `(from, to)`?

File: corehq/apps/app_manager/static/app_manager/js/app_manager.js [hunk 18]
```diff
@@ -326,7 +326,7 @@ hqDefine('app_manager/js/app_manager', function () {
 
             if (to !== from || movingFormToNewModule) {
                 var $form = $sortable.find('> .sort-action form');
-                saveDataToForm($form, from, to, fromModuleUid, toModuleUid, sortingForms);
+                storeRearrangementInForm($form, from, to, fromModuleUid, toModuleUid, sortingForms);
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/app_manager.js]: nice

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/views/apps.py (hunk 17 above)
Comment (esoergel) [reply corehq/apps/app_manager/views/apps.py]: That's fair - I'll add in a docstring.  Nested list comprehensions took me a bit to get, but what finally did it for me is just remembering that the for statements are in the same order as in a regular nested loop.  So these two are equivalent:
```python
players = [player for team in teams for player in team]

players = []
for team in teams:
    for player in team:
        players.add(player)
```
I do think reversing the order would make more sense, but eh, that's how python works, I guess.  Just a matter of getting used to it.  I mentally move `player` to the end.

Comment (orangejenny) [reply corehq/apps/app_manager/views/apps.py]: Sounds good. Thanks for the example - I shall study ;)

File: corehq/apps/app_manager/models.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/app_manager/models.py]: I did also consider switching rearrangement to work by UUID, but figured that'd be a bit too much scope creep and wouldn't really make what I'm working on any easier.  Would be nice to do though - instead of saying "rearrange from position x to position y", say "move module X to immediately after module y".  Then even if someone moved things around after you loaded the page, it'd work properly.

[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: thank you!

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html [hunk 19]
```diff
@@ -100,7 +100,7 @@ <h4 class="modal-title">
                     <button type="button" class="btn btn-default" data-dismiss="modal">
                         {% trans "Not Now" %}
                     </button>
-                    <button type="submit" class="disable-on-submit btn btn-danger">
+                    <button type="submit" class="disable-on-submit btn btn-danger btn-primary">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: remove `btn-danger`

File: corehq/toggles.py [hunk 20]
```diff
@@ -1041,6 +1041,19 @@ def _commtrackify(domain_name, toggle_is_enabled):
     [NAMESPACE_DOMAIN]
 )
 
+LEGACY_CHILD_MODULES = StaticToggle(
```
Comment (orangejenny) [inline corehq/toggles.py]: Should this and the migration be in a separate PR that's released before this one, so the legacy apps are flagged before the new code goes out?

File: corehq/apps/app_manager/migrations/0003_legacy_child_modules.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+from django.db import migrations
+
+from corehq import toggles
+from corehq.apps.app_manager.dbaccessors import get_app_ids_in_domain, get_current_app
+from corehq.util.log import with_progress_bar
+
+
+def flag_legacy_child_module_domains(*args, **kwargs):
+    """Enable the LEGACY_CHILD_MODULES flag for domains that need it"""
+    domains = set(toggles.BASIC_CHILD_MODULE.get_enabled_domains() +
```
Comment (orangejenny) [inline corehq/apps/app_manager/migrations/0003_legacy_child_modules.py]: Should this also include apps that use custom case tiles, even if they don't currently use child modules, in case they turn child modules on in the future?

**Comment by esoergel** [comment]
This PR is getting big, and more work is coming out of QA.  Consequently, I'll no longer be adding commits directly to this PR - instead, I'll PR all changes in from additional branches.

File: corehq/toggles.py (hunk 20 above)
Comment (esoergel) [reply corehq/toggles.py]: Sorry, I thought I'd replied to this days ago.  I wanted the migration be run during the deploy since it also reorders modules in some circumstances, which I'd like to happen right when we roll out the new UI, otherwise it might introduce additional confusion.

File: corehq/apps/app_manager/migrations/0003_legacy_child_modules.py (hunk 21 above)
Comment (esoergel) [reply corehq/apps/app_manager/migrations/0003_legacy_child_modules.py]: I don't think so - the two aren't incompatible, but it's not always backwards compatible.  If a project starts using child modules later, they'll only have an issue when they make a new child module which changes the module indices.  That's similar to adding a normal new module or otherwise changing the ordering, and I think its appropriate to leave it to them to address.

File: corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html (hunk 19 above)
Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/menu/appnav_menu.html]: Good call
https://github.com/dimagi/commcare-hq/pull/23676

File: corehq/toggles.py (hunk 20 above)
Comment (orangejenny) [reply corehq/toggles.py]: sounds good

File: corehq/apps/app_manager/migrations/0003_legacy_child_modules.py (hunk 21 above)
Comment (orangejenny) [reply corehq/apps/app_manager/migrations/0003_legacy_child_modules.py]: cool, thanks for explaining
